File: server/src/main/kotlin/com/github/ntoj/app/server/model/entities/Permission.kt
package com.github.ntoj.app.server.model.entities

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import java.math.BigInteger

@Entity(name = "permission_role")
class PermissionRole(
    @Id @Column(nullable = false) var name: String,
    @Column(nullable = false, columnDefinition = "varchar(1023)") var permission: BigInteger,
)

val BIG_1: BigInteger = BigInteger.ONE

/*
status:

PERM_VIEW                    √
PERM_EDIT_SYSTEM
PERM_SET_PERM                √
PERM_USER_PROFILE            √
PERM_REGISTER_USER           √
PERM_JUDGE                   √
 */
val Permission: Map<String, BigInteger> =
    mapOf(
        // general
        "PERM_VIEW" to (BIG_1 shl 0),
        "PERM_EDIT_SYSTEM" to (BIG_1 shl 1),
        "PERM_SET_PERM" to (BIG_1 shl 2),
        "PERM_USER_PROFILE" to (BIG_1 shl 3),
        "PERM_REGISTER_USER" to (BIG_1 shl 4),
        "PERM_JUDGE" to (BIG_1 shl 5),
        "PERM_EDIT_ANNOUNCEMENT" to (BIG_1 shl 6),
        // problem
        "PERM_PROBLEM_CREATE" to (BIG_1 shl 10),
        "PERM_EDIT_ALL_PROBLEMS" to (BIG_1 shl 11),
        "PERM_EDIT_OWN_PROBLEMS" to (BIG_1 shl 12),
        "PERM_SUBMIT_PROBLEM" to (BIG_1 shl 13),
        "PERM_VIEW_PROBLEMS" to (BIG_1 shl 14),
        "PERM_VIEW_HIDDEN_PROBLEMS" to (BIG_1 shl 15),
        // record
        "PERM_REJUDGE_RECORD" to (BIG_1 shl 20),
        // * TODO: add rejudge problem
        "PERM_REJUDGE_PROBLEM" to (BIG_1 shl 21),
        // article
        "PERM_CREATE_ARTICLE" to (BIG_1 shl 30),
        "PERM_EDIT_ALL_ARTICLES" to (BIG_1 shl 31),
        "PERM_EDIT_OWN_ARTICLES" to (BIG_1 shl 32),
        "PERM_VIEW_ARTICLE" to (BIG_1 shl 33),
        // contest
        "PERM_CREATE_CONTEST" to (BIG_1 shl 40),
        "PERM_EDIT_ALL_CONTESTS" to (BIG_1 shl 41),
        "PERM_EDIT_OWN_CONTESTS" to (BIG_1 shl 42),
        "PERM_ATTEND_CONTEST" to (BIG_1 shl 43),
        // homework
        "PERM_CREATE_HOMEWORK" to (BIG_1 shl 50),
        "PERM_EDIT_HOMEWORK" to (BIG_1 shl 51),
        "PERM_EDIT_OWN_HOMEWORK" to (BIG_1 shl 52),
    )

val PERM_GUEST = Permission["PERM_VIEW"]!!

val PERM_DEFAULT =
    Permission["PERM_VIEW"]!! or
        Permission["PERM_USER_PROFILE"]!! or
        Permission["PERM_EDIT_OWN_PROBLEMS"]!! or
        Permission["PERM_SUBMIT_PROBLEM"]!! or
        Permission["PERM_VIEW_PROBLEMS"]!! or
        Permission["PERM_CREATE_ARTICLE"]!! or
        Permission["PERM_EDIT_OWN_ARTICLES"]!! or
        Permission["PERM_VIEW_ARTICLE"]!! or
        Permission["PERM_EDIT_OWN_CONTESTS"]!! or
        Permission["PERM_ATTEND_CONTEST"]!!

val PERM_ROOT = BigInteger("-1")

File: web/src/lib/Permission.ts
export const PERM = {
  // general
  PERM_VIEW: 1n << 0n,
  PERM_EDIT_SYSTEM: 1n << 1n,
  PERM_SET_PERM: 1n << 2n,
  PERM_USER_PROFILE: 1n << 3n,
  PERM_REGISTER_USER: 1n << 4n,
  PERM_JUDGE: 1n << 5n,
  PERM_EDIT_ANNOUNCEMENT: 1n << 6n,

  // problem
  PERM_PROBLEM_CREATE: 1n << 10n,
  PERM_EDIT_ALL_PROBLEMS: 1n << 11n,
  PERM_EDIT_OWN_PROBLEMS: 1n << 12n,
  PERM_SUBMIT_PROBLEM: 1n << 13n,
  PERM_VIEW_PROBLEMS: 1n << 14n,
  PERM_VIEW_HIDDEN_PROBLEMS: 1n << 15n,

  // record
  PERM_REJUDGE_RECORD: 1n << 20n,
  PERM_REJUDGE_PROBLEM: 1n << 21n,

  // article
  PERM_CREATE_ARTICLE: 1n << 30n,
  PERM_EDIT_ALL_ARTICLES: 1n << 31n,
  PERM_EDIT_OWN_ARTICLES: 1n << 32n,
  PERM_VIEW_ARTICLE: 1n << 33n,

  // contest
  PERM_CREATE_CONTEST: 1n << 40n,
  PERM_EDIT_ALL_CONTESTS: 1n << 41n,
  PERM_EDIT_OWN_CONTESTS: 1n << 42n,
  PERM_ATTEND_CONTEST: 1n << 43n,

  // homework
  PERM_CREATE_HOMEWORK: 1n << 50n,
  PERM_EDIT_HOMEWORK: 1n << 51n,
  PERM_EDIT_OWN_HOMEWORK: 1n << 52n,

  // group
  PERM_CREATE_GROUP: 1n << 60n,
  PERM_EDIT_GROUP: 1n << 61n,
  PERM_EDIT_OWN_GROUP: 1n << 62n,
}

interface Permission {
  family: string
  value: bigint
  desc: string
}

export const Perm = (family: string, value: bigint, desc: string) => ({ family, value, desc } as Permission)

export const PERMS = [
  Perm('general', PERM.PERM_VIEW, 'View system'),
  Perm('general', PERM.PERM_EDIT_SYSTEM, 'Edit system settings'),
  Perm('general', PERM.PERM_SET_PERM, 'Set user role'),
  Perm('general', PERM.PERM_USER_PROFILE, 'User profile'),
  Perm('general', PERM.PERM_REGISTER_USER, 'Register user'),
  Perm('general', PERM.PERM_JUDGE, 'Is user able to sign in as judger'),
  Perm('general', PERM.PERM_EDIT_ANNOUNCEMENT, 'Edit announcement'),

  Perm('problem', PERM.PERM_PROBLEM_CREATE, 'Create problem'),
  Perm('problem', PERM.PERM_EDIT_ALL_PROBLEMS, 'Edit all problems'),
  Perm('problem', PERM.PERM_EDIT_OWN_PROBLEMS, 'Edit own problems'),
  Perm('problem', PERM.PERM_SUBMIT_PROBLEM, 'Submit problem'),
  Perm('problem', PERM.PERM_VIEW_PROBLEMS, 'View problems'),
  Perm('problem', PERM.PERM_VIEW_HIDDEN_PROBLEMS, 'View hidden problems'),

  Perm('record', PERM.PERM_REJUDGE_RECORD, 'Rejudge record'),
  Perm('record', PERM.PERM_REJUDGE_PROBLEM, 'Rejudge problem'),

  Perm('article', PERM.PERM_CREATE_ARTICLE, 'Create article'),
  Perm('article', PERM.PERM_EDIT_ALL_ARTICLES, 'Edit all articles'),
  Perm('article', PERM.PERM_EDIT_OWN_ARTICLES, 'Edit own articles'),
  Perm('article', PERM.PERM_VIEW_ARTICLE, 'View article'),

  Perm('contest', PERM.PERM_CREATE_CONTEST, 'Create contest'),
  Perm('contest', PERM.PERM_EDIT_ALL_CONTESTS, 'Edit all contests'),
  Perm('contest', PERM.PERM_EDIT_OWN_CONTESTS, 'Edit own contests'),
  Perm('contest', PERM.PERM_ATTEND_CONTEST, 'Attend contest'),

  Perm('homework', PERM.PERM_CREATE_HOMEWORK, 'Create homework'),
  Perm('homework', PERM.PERM_EDIT_HOMEWORK, 'Edit homework'),
  Perm('homework', PERM.PERM_EDIT_OWN_HOMEWORK, 'Edit own homework'),

  Perm('group', PERM.PERM_CREATE_GROUP, 'Create group'),
  Perm('group', PERM.PERM_EDIT_GROUP, 'Edit group'),
  Perm('group', PERM.PERM_EDIT_OWN_GROUP, 'Edit own group'),
]

export const PERMS_BY_FAMILY: Record<string, Permission[]> = {}
for (const perm of PERMS) {
  if (!PERMS_BY_FAMILY[perm.family]) {
    PERMS_BY_FAMILY[perm.family] = []
  }
  PERMS_BY_FAMILY[perm.family].push(perm)
}

export function checkPermission(perm: bigint, target: bigint): boolean {
  if (perm === -1n) {
    return true
  }
  return (perm & target) === target
}
